osbs/kerberos_ccache.py

```diff
@@ -0,0 +1,90 @@
+"""
+Copyright (c) 2015 Red Hat, Inc
+All rights reserved.
+
+This software may be modified and distributed under the terms
+of the BSD license. See the LICENSE file for details.
+"""
+from __future__ import print_function, absolute_import, unicode_literals
+
+import re
+import os
+import logging
+import datetime
+import subprocess
+
+from osbs.exceptions import OsbsException
+
+logger = logging.getLogger(__name__)
+
+KLIST_TGT_RE = (r"\d\d/\d\d/\d{2,4}"
+                 " +"
+                 "\d\d:\d\d:\d\d"
+                 " +"
+                 "(?P<month>\d\d)"
+                 "/"
+                 "(?P<day>\d\d)"
+                 "/"
+                 "(?P<year>\d{2,4})"
+                 " +"
+                 "(?P<hour>\d\d)"
+                 ":"
+                 "(?P<minute>\d\d)"
+                 ":"
+                 "(?P<second>\d\d)"
+                 " +"
+                 "krbtgt/(?P<realm>[-.A-Z0-9]+)@(?P=realm)")
+
+def run(cmd, extraenv={}):
+    env = os.environ.copy()
+    env.update(extraenv)
+
+    logger.debug("Subprocess: %s", ' '.join(cmd))
+    # universal_newlines=True causes stdout/stderr to be strings (not bytes) in py3
+    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, env=env,
+                         universal_newlines=True)
+    stdout, stderr = p.communicate()
+
+    return p.returncode, stdout, stderr
+
+def kerberos_ccache_init(principal, keytab_file, ccache_file=None):
+    """
+    Checks whether kerberos credential cache has ticket-granting ticket that is valid for at least
+    an hour.
+
+    Default ccache is used unless ccache_file is provided. In that case, KRB5CCNAME environment
+    variable is set to the value of ccache_file if we successfully obtain the ticket.
+    """
+    tgt_valid = False
+    env = { "LC_ALL": "C" } # klist uses locales to format date on RHEL7+
```

Comment:
So, klist tries a variety of formats until it finds one that fits within the width it wants. :-/